### PR TITLE
[DROOLS-7493] [DROOLS-7497] executable model wrongly rewrites modify in if-block

### PR DESCRIPTION
**Pull request by tkobayas**
- Additional tests to cover setter order for properperty reactivity [DROOLS-7497]

**Ports** 
This is a PR for main
for 7.x ? ->

**JIRA**:
https://issues.redhat.com/browse/DROOLS-7493
https://issues.redhat.com/browse/DROOLS-7497

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] tests</b>

- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

- <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

- <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-lts</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native</b>

 - for <b>native lts checks</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins run native-lts</b>

- for a <b>specific native lts check</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native-lts</b>
</details>

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

**Comment by tkobayas** [comment]
Hi @mariofusco  @lucamolteni 

This PR is still draft and has several test failures.

Firstly, this PR solves the issue in https://issues.redhat.com/browse/DROOLS-7493

A) In exec-model, `modify` is always rendered as `drools.update` at the end of consequence, so for example, it is moved out from if-block.
B) In exec-model, if `modify` block doesn't have setters,  `drools.update` is not rendered even if setters exist before the `modify` block (it works with non-exec-model). 

So I explicitly add `drools.update` in the same block where `modify` existed.

Then, another concern arises. Do we want to enlist properties for property reactivity which are set *after* `modify`/`update`?

e.g. In this case, `value2` should be enlisted as property reactivity? (https://issues.redhat.com/browse/DROOLS-7497)
```
modify ($fact) {
  setValue1(2)
}
$fact.setValue2(2);
```

C) In non-exec-model, 
C.1) With java dialect, RHS is analyzed only before `modify`/`update`. So `value2` is not property reactive
C.2) With mvel dialect, the whole RHS is analyzed. So `value2` is property reactive
D) In exec-model, with both dialect, RHS is analyzed only before `modify`/`update`. So `value2` is not property reactive. *However*, as described above, the current implementation (before this PR) renders `modify` as `drools.update` at the end of consequence. So `modify` enlists all setters in RHS.

So, more or less, there is inconsistency in the behavior.

Personaly, I like C.2) behavior (= analyze whole RHS and make them all property reactive), because `PropagationEntry.Update` is evaluated after whole RHS execution regardless of the `modify` block position in the RHS, so anyway, all changes in RHS will affects in the evaluation.

Could you share your thoughts on this? Thanks!

(Sorry, in my `PropertyReactivityMatrixTest`, `value1` and `value2` order is inverted, in order to meet with existing rules)

**Comment by kie-ci4** [comment]
**(tests) - drools job** `#1529` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>


export BUILD_MVN_OPTS_CURRENT=-Dfull -Prun-code-coverage
build-chain build cross_pr -f 'https://raw.githubusercontent.com/kiegroup/drools/main/.ci/buildchain-config.yaml' -o 'bc' -p kiegroup/drools -u https://github.com/kiegroup/drools/pull/5380 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1529/display/redirect

**Test results:**
- PASSED: 21637
- FAILED: 8

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1529/testReport/org.drools.model.codegen.execmodel/PropertyReactivityMatrixTest/assignmentBeforeAndAfterModify_shouldTriggerReactivity[STANDARD_FROM_DRL_JAVA]/">org.drools.model.codegen.execmodel.PropertyReactivityMatrixTest.assignmentBeforeAndAfterModify_shouldTriggerReactivity[STANDARD_FROM_DRL JAVA]</a></summary>
[Should trigger property reactivity on value1] <br/>Expecting actual:<br/>  []<br/>to contain exactly (and in same order):<br/>  ["R2 fired"]<br/>but could not find the following elements:<br/>  ["R2 fired"]<br/>
</details>
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1529/testReport/org.drools.model.codegen.execmodel/PropertyReactivityMatrixTest/assignmentBeforeAndAfterUpdate_shouldTriggerReactivity[STANDARD_FROM_DRL_JAVA]/">org.drools.model.codegen.execmodel.PropertyReactivityMatrixTest.assignmentBeforeAndAfterUpdate_shouldTriggerReactivity[STANDARD_FROM_DRL JAVA]</a></summary>
[Should trigger property reactivity on value1] <br/>Expecting actual:<br/>  []<br/>to contain exactly (and in same order):<br/>  ["R2 fired"]<br/>but could not find the following elements:<br/>  ["R2 fired"]<br/>
</details>
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1529/testReport/org.drools.model.codegen.execmodel/PropertyReactivityMatrixTest/assignmentAfterIfBlockModify_shouldTriggerReactivity[STANDARD_FROM_DRL_JAVA]/">org.drools.model.codegen.execmodel.PropertyReactivityMatrixTest.assignmentAfterIfBlockModify_shouldTriggerReactivity[STANDARD_FROM_DRL JAVA]</a></summary>
[Should trigger property reactivity on value1] <br/>Expecting actual:<br/>  []<br/>to contain exactly (and in same order):<br/>  ["R2 fired"]<br/>but could not find the following elements:<br/>  ["R2 fired"]<br/>
</details>
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1529/testReport/org.drools.model.codegen.execmodel/PropertyReactivityMatrixTest/assignmentAfterModify_shouldTriggerReactivity[STANDARD_FROM_DRL_JAVA]/">org.drools.model.codegen.execmodel.PropertyReactivityMatrixTest.assignmentAfterModify_shouldTriggerReactivity[STANDARD_FROM_DRL JAVA]</a></summary>
[Should trigger property reactivity on value1] <br/>Expecting actual:<br/>  []<br/>to contain exactly (and in same order):<br/>  ["R2 fired"]<br/>but could not find the following elements:<br/>  ["R2 fired"]<br/>
</details>
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1529/testReport/org.drools.model.codegen.execmodel/PropertyReactivityMatrixTest/assignmentBeforeAndAfterUpdate_shouldTriggerReactivity[PATTERN_DSL_JAVA]/">org.drools.model.codegen.execmodel.PropertyReactivityMatrixTest.assignmentBeforeAndAfterUpdate_shouldTriggerReactivity[PATTERN_DSL JAVA]</a></summary>
[Should trigger property reactivity on value1] <br/>Expecting actual:<br/>  []<br/>to contain exactly (and in same order):<br/>  ["R2 fired"]<br/>but could not find the following elements:<br/>  ["R2 fired"]<br/>
</details>
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1529/testReport/org.drools.model.codegen.execmodel/PropertyReactivityMatrixTest/assignmentAfterIfBlockModify_shouldTriggerReactivity[PATTERN_DSL_JAVA]/">org.drools.model.codegen.execmodel.PropertyReactivityMatrixTest.assignmentAfterIfBlockModify_shouldTriggerReactivity[PATTERN_DSL JAVA]</a></summary>
[Should trigger property reactivity on value1] <br/>Expecting actual:<br/>  []<br/>to contain exactly (and in same order):<br/>  ["R2 fired"]<br/>but could not find the following elements:<br/>  ["R2 fired"]<br/>
</details>
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1529/testReport/org.drools.model.codegen.execmodel/PropertyReactivityMatrixTest/assignmentBeforeAndAfterUpdate_shouldTriggerReactivity[PATTERN_DSL_MVEL]/">org.drools.model.codegen.execmodel.PropertyReactivityMatrixTest.assignmentBeforeAndAfterUpdate_shouldTriggerReactivity[PATTERN_DSL MVEL]</a></summary>
[Should trigger property reactivity on value1] <br/>Expecting actual:<br/>  []<br/>to contain exactly (and in same order):<br/>  ["R2 fired"]<br/>but could not find the following elements:<br/>  ["R2 fired"]<br/>
</details>
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1529/testReport/org.drools.model.codegen.execmodel/PropertyReactivityMatrixTest/assignmentAfterIfBlockModify_shouldTriggerReactivity[PATTERN_DSL_MVEL]/">org.drools.model.codegen.execmodel.PropertyReactivityMatrixTest.assignmentAfterIfBlockModify_shouldTriggerReactivity[PATTERN_DSL MVEL]</a></summary>
[Should trigger property reactivity on value1] <br/>Expecting actual:<br/>  []<br/>to contain exactly (and in same order):<br/>  ["R2 fired"]<br/>but could not find the following elements:<br/>  ["R2 fired"]<br/>
</details>

**Comment by lucamolteni** [comment]
> Personaly, I like C.2) behavior (= analyze whole RHS and make them all property reactive), because `PropagationEntry.Update` is evaluated after whole RHS execution regardless of the `modify` block position in the RHS, so anyway, all changes in RHS will affects in the evaluation.
> 
> Could you share your thoughts on this? Thanks!
> 
> (Sorry, in my `PropertyReactivityMatrixTest`, `value1` and `value2` order is inverted, in order to meet with existing rules)

I agree there should be some consistency eventually, and C.2 seems the good choice. 

We have considered using the mvel-compiler for MVEL and Java "classic" (non exec model) evaluation as well so once we solve it there it could be possible to have it everywhere, eventually

**Comment by tkobayas** [comment]
> We have considered using the mvel-compiler for MVEL and Java "classic" (non exec model) evaluation as well so once we solve it there it could be possible to have it everywhere, eventually

Great. It should be nice :)

**Comment by mariofusco** [comment]
This not an obvious choice and deserves a few considerations. First of all, let me do a small digression, starting from user's expectations (that unfortunately cannot be fulfilled and I will explain why). If I write a consequence like:

```
modify ($fact) {
  setValue1(2)
}
$fact.setValue2(2);
```

I would expect that it would be syntactic sugar for something like:

```
$fact.setValue1(2);
update($fact);
$fact.setValue2(2);
```

meaning that I'd expect the instruction to be executed in that order. First I set `value1` and propagate the update of the fact into the engine with only that change and only after I also change `value2`, but without notifying the engine of this second modification. 

As anticipated however, given the current implementation, this expectation cannot be satisfied because what the `update` does is only scheduling an action on the `propagation list`, and when that update action is dequeued and actually executed the fact to be updated also has `value2` already changed.

Long story short, ideally I'd prefer C.1, but since it doesn't reflect how the fact update is propagated into the engine we are practically obliged to go with C.2. More importantly we need to make sure that the C.2 behavior is enforced consistently with and without exec model and regardless of the dialect.

**Comment by tkobayas** [comment]
> Long story short, ideally I'd prefer C.1, but since it doesn't reflect how the fact update is propagated into the engine we are practically obliged to go with C.2. More importantly we need to make sure that the C.2 behavior is enforced consistently with and without exec model and regardless of the dialect.

Thanks, I completely agree with you. I will also add docs to explain the behavior.

**Comment by kie-ci4** [comment]
**(tests) - drools job** `#1532` was: **FAILURE**
Possible explanation: Pipeline failure or project build failure


<details>
<summary><b>Reproducer</b></summary>


export BUILD_MVN_OPTS_CURRENT=-Dfull -Prun-code-coverage
build-chain build cross_pr -f 'https://raw.githubusercontent.com/kiegroup/drools/main/.ci/buildchain-config.yaml' -o 'bc' -p kiegroup/drools -u https://github.com/kiegroup/drools/pull/5380 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1532/display/redirect

**Test results:**
- PASSED: 21655
- FAILED: 0

Those are the test failures: none

See console log:
<details>
<summary><b>Console Logs</b></summary>
[2023-07-11T09:17:30.832Z]     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)<br/>[2023-07-11T09:17:30.832Z]     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)<br/>[2023-07-11T09:17:30.832Z]     at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)<br/>[2023-07-11T09:17:30.832Z]     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)<br/>[2023-07-11T09:17:30.832Z]     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)<br/>[2023-07-11T09:17:30.832Z]     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)<br/>[2023-07-11T09:17:30.832Z]     at java.lang.reflect.Method.invoke (Method.java:566)<br/>[2023-07-11T09:17:30.832Z]     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)<br/>[2023-07-11T09:17:30.832Z]     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)<br/>[2023-07-11T09:17:30.832Z]     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)<br/>[2023-07-11T09:17:30.832Z]     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)<br/>[2023-07-11T09:17:30.832Z] [ERROR] <br/>[2023-07-11T09:17:30.832Z] [ERROR] Re-run Maven using the -X switch to enable full debug logging.<br/>[2023-07-11T09:17:30.832Z] [ERROR] <br/>[2023-07-11T09:17:30.832Z] [ERROR] For more information about the errors and possible solutions, please read the following articles:<br/>[2023-07-11T09:17:30.832Z] [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException<br/>[2023-07-11T09:17:30.832Z] [ERROR] <br/>[2023-07-11T09:17:30.832Z] [ERROR] After correcting the problems, you can resume the build with the command<br/>[2023-07-11T09:17:30.832Z] [ERROR]   mvn <args> -rf :drools-drlonyaml-cli-tests<br/>[2023-07-11T09:17:31.086Z] + test 1 -eq 0<br/>[Pipeline] }<br/>[Pipeline] // withCredentials<br/>[Pipeline] }<br/>[Pipeline] // dir<br/>[Pipeline] }<br/>[Pipeline] // script<br/>[Pipeline] }<br/>[Pipeline] // withEnv<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Upload artifacts to repository)<br/>Stage "Upload artifacts to repository" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (check space after build)<br/>Stage "check space after build" skipped due to earlier failure(s)<br/>[Pipeline] }<br/>[Pipeline] // stage<br/>[Pipeline] stage<br/>[Pipeline] { (Declarative: Post Actions)<br/>[Pipeline] script<br/>[Pipeline] {<br/>[Pipeline] sh<br/>[2023-07-11T09:17:32.028Z] + wget --no-check-certificate -qO - https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1532/api/json<br/>[Pipeline] readJSON<br/>[Pipeline] sh<br/>[2023-07-11T09:17:32.349Z] + wget --no-check-certificate -qO - https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.drools/1532/consoleText<br/>[2023-07-11T09:17:32.349Z] + tail -n 50<br/>
</details>

**Review by mariofusco — APPROVED** [review]

**Review by lucamolteni — APPROVED** [review]

**Review by pibizza — APPROVED** [review]
Looks good to me.

**Comment by tkobayas** [comment]
@mariofusco please merge this PR, thanks!
